# Consigna del desafío 1

**1**. Vectorizar documentos. Tomar 5 documentos al azar y medir similaridad con el resto de los documentos.
Estudiar los 5 documentos más similares de cada uno analizar si tiene sentido
la similaridad según el contenido del texto y la etiqueta de clasificación.

**2**. Entrenar modelos de clasificación Naïve Bayes para maximizar el desempeño de clasificación
(f1-score macro) en el conjunto de datos de test. Considerar cambiar parámteros
de instanciación del vectorizador y los modelos y probar modelos de Naïve Bayes Multinomial
y ComplementNB.

**3**. Transponer la matriz documento-término. De esa manera se obtiene una matriz
término-documento que puede ser interpretada como una colección de vectorización de palabras.
Estudiar ahora similaridad entre palabras tomando 5 palabras y estudiando sus 5 más similares. **La elección de palabras no debe ser al azar para evitar la aparición de términos poco interpretables, elegirlas "manualmente"**.

## Imports y preparación dataset

In [6]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.naive_bayes import MultinomialNB, ComplementNB
from sklearn.metrics import f1_score
from sklearn.datasets import fetch_20newsgroups
import numpy as np
import random

In [2]:
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))
display(newsgroups_test.target_names)

tfidfvect = TfidfVectorizer()
X_train = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target

['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']

### Ejemplo similitud de documentos 

In [ ]:
idx = 4811
display(newsgroups_train.data[idx])
cossim = cosine_similarity(X_train[idx], X_train)[0]
np.sort(cossim)[::-1]       #Valores de similaridad ordenados de mayor a menor
np.argsort(cossim)[::-1]    #Indices de documentos similares de mayor a menor
mostsim = np.argsort(cossim)[::-1][1:6] #Top 5
for i in mostsim:
  display(newsgroups_train.target_names[y_train[i]])

"THE WHITE HOUSE\n\n                  Office of the Press Secretary\n                   (Pittsburgh, Pennslyvania)\n______________________________________________________________\nFor Immediate Release                         April 17, 1993     \n\n             \n                  RADIO ADDRESS TO THE NATION \n                        BY THE PRESIDENT\n             \n                Pittsburgh International Airport\n                    Pittsburgh, Pennsylvania\n             \n             \n10:06 A.M. EDT\n             \n             \n             THE PRESIDENT:  Good morning.  My voice is coming to\nyou this morning through the facilities of the oldest radio\nstation in America, KDKA in Pittsburgh.  I'm visiting the city to\nmeet personally with citizens here to discuss my plans for jobs,\nhealth care and the economy.  But I wanted first to do my weekly\nbroadcast with the American people. \n             \n             I'm told this station first broadcast in 1920 when\nit reported th

'talk.politics.misc'

'talk.politics.misc'

'talk.politics.misc'

'talk.politics.misc'

'talk.politics.misc'

### Ejemplo clasificación

In [4]:
clf = MultinomialNB()
clf.fit(X_train, y_train)
X_test = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target
y_pred =  clf.predict(X_test)
display(f"F1 score {f1_score(y_test, y_pred, average='macro')}")

'F1 score 0.5854345727938506'

## Desarrollo 

### Punto 1

In [81]:
def get_docs_cosine_similarity(X_train, idx, n):
    topn_dict = {"doc_idx": idx}
    cossim = cosine_similarity(X_train[idx], X_train)[0]    
    topn_dict.update({"docs_sim_idx" : np.argsort(cossim)[::-1][1:n+1], "docs_sim_cosin" : np.sort(cossim)[::-1][1:n+1]})
    return topn_dict

def print_doc_info(dataset, topn_docs):
    for docs in topn_docs:
        text = dataset.data[docs['doc_idx']]
        type = dataset.target_names[int(dataset.target[docs['doc_idx']])]
        display(f"Doc referencia ID:{docs['doc_idx']}, clase: {type}")
        display(f"CONTENIDO: {text}")
        for i in range(len(docs['docs_sim_idx'])):
            text = dataset.data[docs['docs_sim_idx'][i]]
            type = dataset.target_names[int(dataset.target[docs['docs_sim_idx'][i]])]
            display(f"{i+1}) Doc similar ID: {docs['docs_sim_idx'][i]}, similitud coseno: {docs['docs_sim_cosin'][i]}, clase: {type}")
            display(f"CONTENIDO: {text}")
        display('-'*150)

In [82]:
seed = 42
random.seed(seed)
n_docs = 5
sample_docs_idx = random.sample(range(X_train.shape[0]), n_docs)

topn_docs = []
for i in sample_docs_idx:
    topn_docs.append(get_docs_cosine_similarity(X_train, i, 5))
print_doc_info(newsgroups_train, topn_docs)


'Doc referencia ID:10476, clase: rec.sport.hockey'

"CONTENIDO: This is a general question for US readers:\n\nHow extensive is the playoff coverage down there?  In Canada, it is almost\nimpossible not to watch a series on TV (ie the only two series I have not had\nan opportunity to watch this year are Wash-NYI and Chi-Stl, the latter because\nI'm in the wrong time zone!).  We (in Canada) are basically swamped with \ncoverage, and I wonder how many series/games are televised nationally or even\nlocally in the US and how much precedence they take over, say, local news if\nthe games go into double-OT.\n\nEmail me so as not to waste bandwidth, please.  My news feed is kind of slow\nanyways."

'1) Doc similar ID: 5064, similitud coseno: 0.22503645427838304, clase: rec.sport.hockey'

"CONTENIDO: \nI only have one comment on this:  You call this a *classic* playoff year\nand yet you don't include a Chicago-Detroit series.  C'mon, I'm a Boston\nfan and I even realize that Chicago-Detroit games are THE most exciting\ngames to watch."

'2) Doc similar ID: 9623, similitud coseno: 0.21743229780919432, clase: talk.politics.mideast'

'CONTENIDO: Accounts of Anti-Armenian Human Right Violations in Azerbaijan #012\n                 Prelude to Current Events in Nagorno-Karabakh\n\n        +---------------------------------------------------------+\n        |                                                         |\n        |  I saw a naked girl with her hair down. They were       |\n        |  dragging her. She kept falling because they were       |\n        |  pushing her and kicking her. She fell down, it was     |\n        |  muddy there, and later other witnesses who saw it from |\n        |  their balconies told us, they seized her by the hair   |\n        |  and dragged her a couple of blocks, as far as the      |\n        |  mortgage bank, that\'s a good block and a half or two   |\n        |  from here. I know this for sure because I saw it       |\n        |  myself.                                                |\n        |                                                         |\n        +---------------

'3) Doc similar ID: 10575, similitud coseno: 0.21644447761221192, clase: sci.crypt'

'CONTENIDO: \nI am not an expert in the cryptography science, but some basic things\nseem evident to me, things which this Clinton Clipper do not address.\nThe all pertain to opportunites for abuse, and conclusions based on what\nI have seen the membership of this group (except for two notable persons)\nagree to.  If anything bad is possible by the government in theory, it\nalmost always ends up happening in fact.  So the key is to make abuse\nIMPOSSIBLE.  Question authority, and remember power DOES corrupt.\n\nI think one has to regard this whole idea in the sense that it WILL be\nexpanded to include other data forms, such as data transfer and stored\ndata. and as such should be treated as if it were expanded (or such\nexpansion will be almost impossible to stop, using Clipper as a precident).\nThere was a hint of that in the proposal, remember.  That said, please\nbear with me, I am not very articulate, so I take more words to say\nwhat others could say much more briefly.  Thank you.

'4) Doc similar ID: 10836, similitud coseno: 0.21260319521108376, clase: alt.atheism'

'CONTENIDO: Archive-name: atheism/faq\nAlt-atheism-archive-name: faq\nLast-modified: 5 April 1993\nVersion: 1.1\n\n                    Alt.Atheism Frequently-Asked Questions\n\nThis file contains responses to articles which occur repeatedly in\nalt.atheism.  Points covered here are ones which are not covered in the\n"Introduction to Atheism"; you are advised to read that article as well\nbefore posting.\n\nThese answers are not intended to be exhaustive or definitive. The purpose of\nthe periodic FAQ postings is not to stifle debate, but to raise its level. If\nyou have something to say concerning one of these questions and which isn\'t\ncovered by the answer given, please feel free to make your point.\n\nOverview of contents:\n\n   "What is the purpose of this newsgroup?"\n   "Hitler was an atheist!"\n   "The Bible proves it"\n   "Pascal\'s Wager"\n   "What is Occam\'s Razor?"\n   "Why it\'s good to believe in Jesus"\n   "Why I know that God exists"\n   "Einstein and "God does not pla

'5) Doc similar ID: 2350, similitud coseno: 0.21107117783786433, clase: sci.crypt'

"CONTENIDO: Archive-name: net-privacy/part1\nLast-modified: 1993/3/3\nVersion: 2.1\n\n\nIDENTITY, PRIVACY, and ANONYMITY on the INTERNET\n================================================\n\n(c) 1993 L. Detweiler.  Not for commercial use except by permission\nfrom author, otherwise may be freely copied.  Not to be altered. \nPlease credit if quoted.\n\nSUMMARY\n=======\n\nInformation on email and account privacy, anonymous mailing and \nposting, encryption, and other privacy and rights issues associated\nwith use of the Internet and global networks in general.\n\n(Search for <#.#> for exact section. Search for '_' (underline) for\nnext section.)\n\nPART 1\n====== (this file)\n\nIdentity\n--------\n<1.1> What is `identity' on the internet?\n<1.2> Why is identity (un)important on the internet?\n<1.3> How does my email address (not) identify me and my background?\n<1.4> How can I find out more about somebody from their email address?\n<1.5> Why is identification (un)stable on the internet?

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Doc referencia ID:1824, clase: comp.sys.mac.hardware'

'CONTENIDO: \n\n\tI think this kind of comparison is pretty useless in general.  The\nprocessor is only good when a good computer is designed around it adn the\ncomputer is used in its designed purpose.  Comparing processor speed is\npretty dumb because all you have to do is just increase the clock speed\nto increase speed among other things.\n\n\tI mean how can you say a 040 is faster than a 486 without \ngiving is operational conditions?  Can you say the same when \nyou are running a program that uses a lot of transidental functions.\nKnowing that 040 does not have transidental functions building in to \nits FPU and 486 does, can you say that 040 is still faster?\n\n\tAnyway, I hope people do not decided upon wether a computers\nis good or not solely on its processor.  Or how fast a processor is\nbased on its name, because one can alway do a certain things to a\nprocessor to speed it up.  \n\n\tBut if we restrict our arguements to, for example, pure\nprocessor architectural issues.  

'1) Doc similar ID: 9921, similitud coseno: 0.35419792160817193, clase: comp.sys.mac.hardware'

'CONTENIDO: dhk@ubbpc.uucp (Dave Kitabjian) writes ...\n\n040 486 030 386 020 286\n\n\n060 fastest, then Pentium, with the first versions of the PowerPC\nsomewhere in the vicinity.\n\n\nNo.  Computer speed is only partly dependent of processor/clock speed.\nMemory system speed play a large role as does video system speed and\nI/O speed.  As processor clock rates go up, the speed of the memory\nsystem becomes the greatest factor in the overall system speed.  If\nyou have a 50MHz processor, it can be reading another word from memory\nevery 20ns.  Sure, you can put all 20ns memory in your computer, but\nit will cost 10 times as much as the slower 80ns SIMMs.\n\nAnd roughly, the 68040 is twice as fast at a given clock\nspeed as is the 68030.\n'

'2) Doc similar ID: 6364, similitud coseno: 0.31319104088093985, clase: comp.sys.mac.hardware'

'CONTENIDO: Well folks, after some thought the answer struck me flat in the face:\n\n"Why would Apple release a Duo Dock with a processor of its own?"\n\nHere\'s why- People have hounded Apple for a notebook with a 68040 processor\nin it. Apple can\'t deliver that right now because the 040 saps too much\npower, radiates far too much heat, and is too large for a notebook. How\ndoes one get around that without designing a new chipset? Use existing\nPowerBook technology to your best advantage. The Duo Dock gives Apple a\nunique ability to give users that 040 power in a "Semi-Portable" fashion.\nBy plunking the 040 into the Dock, you\'ve got "quadra" power at your desk.\nOn the road, that 33mhz 68030 should be able to handle most of your needs.\nOkay, not the BEST solution, but its an answer to a no-win situation. :-)\nSo, does this mean one will be able to use the PowerBook\'s processor in\nparallel to the dock\'s processor? Okay, we\'re getting REALLY hypothetical\nnow... '

'3) Doc similar ID: 5509, similitud coseno: 0.30413670525282677, clase: comp.sys.mac.hardware'

'CONTENIDO: rvenkate@ux4.cso.uiuc.edu (Ravikuma Venkateswar) writes ...\n\nBenchmarks are for marketing dweebs and CPU envy.  OK, if it will make\nyou happy, the 486 is faster than the 040.  BFD.  Both architectures\nare nearing then end of their lifetimes.  And especially with the x86\narchitecture: good riddance.\n\n\nThe point being the processor speed is only one of many aspects of a\ncomputers performance.  Clock speed, processor, memory speed, CPU\narchitecture, I/O systems, even the application program all contribute \nto the overall system performance.\n\n\nLook them up yourself.\n'

'4) Doc similar ID: 2641, similitud coseno: 0.2503826782386589, clase: comp.sys.mac.hardware'

"CONTENIDO: \n    I think this is mostly the fault of the people who write up the\nliterature and price lists being confused themselves. Since there are\ntwo possible processor configurations and one of the them doesn't have\nan FPU it does seem to be an option, even though it really isn't.\n\n\n    Well, then allow me to end your confusion. The C650 ONLY come with\nan LC040 in the base 4/80 configuration. If you are not getting this\nconfiguration then you are getting an FPU.\n\n\n    Good question. I have been wondering that since Feb. 10th.\n\n \n    This is possible, but an option is something that you are supposed\nto be able to request when you want it. What Apple has done is given the\nbuyer a CHOICE between configurations and not an OPTION.\n\n\n    This is not unclear at all. In fact Apple has included in the ROMs\nof those machines with LC040s code to recognize the presence of the full\n040's FPU and use it. Thereby making the upgrade as easy as switching\nchips. You pop the 

'5) Doc similar ID: 4359, similitud coseno: 0.24167619736986773, clase: comp.sys.mac.hardware'

'CONTENIDO: If you get the Centris 650 with CD configuration, you are getting a Mac with\na 68RC040 processor that has built-in math coprocessor support.  My \nunderstanding is that the "optional fpu" refers to your option of purchasing\nthe Centris 650 4/80 without FPU OR one of the other configurations WITH FPU.\n\nApple does not offer an upgrade from the non-FPU system to become an FPU\nsystem.  And, it is unclear whether the \'040 processor on the non-FPU system\n(a 68LC040) can be replaced with a 68RC040 supplied by another vendor.\nApple did send a memo out at one point sating that the Centris 610, which ONLY\ncomes with a non-FPU 68LC040 processor CANNOT be upgraded to support an FPU -\nthe pin configurations of the two chips apparently do not match so you cannot\nswap one for another (again, according to Apple\'s memo).\n\nHope that helps.'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Doc referencia ID:409, clase: comp.graphics'

"CONTENIDO: I can't fiqure this out.  I have properly compiled pov on a unix machine\nrunning SunOS 4.1.3  The problem is that when I run the sample .pov files and\nuse the EXACT same parameters when compiling different .tga outputs.  Some\nof the .tga's are okay, and other's are unrecognizable by any software."

'1) Doc similar ID: 3444, similitud coseno: 0.23053284689766265, clase: comp.graphics'

"CONTENIDO: Hi, I'm just getting into PoVRay and I was wondering if there is a graphic\npackage that outputs .POV files.  Any help would be appreciated.\nThanks.\n\nLater'ish\nCraig\n"

'2) Doc similar ID: 5799, similitud coseno: 0.20907273908737237, clase: comp.graphics'

'CONTENIDO: I finally got a 24 bit viewer for my POVRAY generated .TGA files.\n\nIt was written in C by Sean Malloy and he kindly sent me a copy.  He\nwrote it for the same purpose, to view .TGA files using his SpeedStar 24.\n\nIt ONLY works with the SpeedStar 24 and I cannot send copies since it is\nnot my program.  I believe the author may release a version at a future\ntime when the program is more developed.   He may or may not comment on\nthis, as he pleases.\n\nThanks to all who were helpful.\n\nRegards,'

'3) Doc similar ID: 5905, similitud coseno: 0.1981754261808808, clase: comp.graphics'

"CONTENIDO: \nHallo POV-Renderers !\nI've got a BocaX3 Card. Now I try to get POV displaying True Colors\nwhile rendering. I've tried most of the options and UNIVESA-Driver\nbut what happens isn't correct.\nCan anybody help me ?\n"

'4) Doc similar ID: 1764, similitud coseno: 0.18384984554054173, clase: comp.graphics'

"CONTENIDO: hi guys\n like all people in this group i'm a fans of fractal and render sw\n my favourite are fractint pov & 3dstudio 2.0 \n now listen my ideas\n i'have just starting now to be able to use 3dstudio quite well\n so i'm simulating a full animation of a f1 grand prix\n unfortanatly just some lap(10?)\n i' m very interested about all kind of .prj .3ds and so on\n concerning about cars or parts of its (motors wheel ...)\n (dxf are good enough)\n does anyone have object to give me to complete my hard animation\n\n\n anyway any exchanges about object material project will\n be VERY APRECIATE!!!!!\n\n is there a ftp site where I can find its?\n\n i' m looking for .pov files too\n (i 'm interested about cpu time comparision rendering images on\n pov & 3dstusio)\n\n thank to all\n"

'5) Doc similar ID: 3364, similitud coseno: 0.16585392152558243, clase: comp.graphics'

'CONTENIDO: Does anyone know of a good way (standard PC application/PD utility) to\nconvert tif/img/tga files into LaserJet III format.  We would also like to\ndo the same, converting to HPGL (HP plotter) files.\n\nPlease email any response.\n\nIs this the correct group?\n\nThanks in advance.  Michael.'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Doc referencia ID:4506, clase: rec.autos'

"CONTENIDO: \nThis does sound good, but I heard it tends to leave more grit, etc in the \noil pan.  Also, I've been told to change the old when it's hot before the\ngrit has much time to settle.\n\nAny opinions?\n"

'1) Doc similar ID: 4211, similitud coseno: 0.18936063539057416, clase: rec.motorcycles'

"CONTENIDO: \n\nIt's normal for the BMW K bikes to use a little oil in the first few thousand \nmiles.  I don't know why.  I've had three new K bikes, and all three used a\nbit of oil when new - max maybe .4 quart in first 1000 miles; this soon quits\nand by the time I had 10,000 miles on them the oil consumption was about zero.\nI've been told that the harder you run the bike (within reason) the sooner\nit stops using any oil.\n"

'2) Doc similar ID: 5928, similitud coseno: 0.16823674420892193, clase: comp.sys.mac.hardware'

'CONTENIDO: or\nthere\n\n\nOkay, I guess its time for a quick explanation of Mac sound.\n\nThe original documentation for the sound hardware (IM-3) documents how to\nmake sound by directly accessing hardware.  Basically, you jam values\ninto all the even bytes from SoundBase to SoundBase+0x170. This was\nbecause\nof how the Mac 128 (and some later machines) generated sound was by\nscanning\nthis block and D/Aing every even byte (the odd bytes went to the floppy on\nsome machines).\n\nWhen the Mac II (and Apple Sound Chip) was invented, it was designed to\ngenerate stereo sound. It was also designed to be compatible with we had\nonce\ndocumented. So storing bytes at the even values at SoundBase meant "I\nwant to\nplay a mono sound" and so it was emulated. But Apple had since retracted\nthe\ndocumentation on SoundBase and decided not to document the lowest layers\nof\nsound generation. So apple never explained where to stuff bytes if you\nwant\nto make stereo sound.\n\nEnter game develop

'3) Doc similar ID: 6224, similitud coseno: 0.1583340724404407, clase: rec.autos'

"CONTENIDO: Archive-name: rec-autos/part5\n\n[this article is one of a pair of articles containing commonly\nasked automotive questions; the other article contains questions\nmore geared to the automotive enthusiast, and so is not crossposted\nto misc.consumers.  -- rpw]\n\n   [changes as of 14 April 1993: revised brake fluid section, as\n    non-silicone DOT-5 fluids are now apparently available -- rpw]\n\n                 Commonly Asked Automotive Questions\n\n\nTire Questions:\n\nQ:  What do the funny numbers on the sides of my tires mean?\n\nA:  Typically, you will see something like 195/60HR14.  the 195 is the\n    overall width of the tire in millimeters, the tread is usually\n    narrower.  The 60 is the `aspect ratio'; it indicates the height of the\n    sidewall of the tire relative to the overall width.  Our example tire\n    has a sidewall height of 0.60 * 195 ~= 117 mm.  The 14 is the wheel\n    diameter in inches; there are also some special tires called `TRX'\n    tires w

'4) Doc similar ID: 5171, similitud coseno: 0.157693767056016, clase: rec.autos'

"CONTENIDO: \nIf the tire has a leak you should fix it. \n\n\nDoesn't work too well if the engine is hot, its more accurate to check the\noil when the engine is cool, i.e. not when you are at a gas station."

'5) Doc similar ID: 9491, similitud coseno: 0.1521899666815704, clase: rec.autos'

'CONTENIDO: My friend brought a subaru SVX recently.  I had drove it for couples times and I\nthink its a great car, esp on snow.  However when she took it to a local Subaru\ndealer for a oil change, the bill came out to be about 80 dollars.  The dealer\ntold us it is because to change the oil filter on a SVX it is necessary to\ndisassemble a metal cover under the engine and that took an hour of labour.\nAt first, we think we are being ripped off so she phone to a dealer in Toronto\nbut found out the they are charging roughly the same price.  So is there any\nSVX owner out there that has the same problem ?  And if the oil change story is\ntrue, then the engineer of Subaru looks pretty stubid to me. By the way, the car\nlooks great.'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Doc referencia ID:4012, clase: rec.sport.hockey'

"CONTENIDO: For those Leaf fans who are concerned, the following players are slated for\nreturn on Thursday's Winnipeg-Toronto game :\n    Peter Zezel, John Cullen\n\n  Mark Osborne and Dave Ellett are questionable to return on Thursday."

'1) Doc similar ID: 6599, similitud coseno: 0.16001317417336078, clase: soc.religion.christian'

'CONTENIDO: True.\n\nAlso read 2 Peter 3:16\n\nPeter warns that the scriptures are often hard to understand by those who\nare not learned on the subject.'

'2) Doc similar ID: 10644, similitud coseno: 0.14276898565096252, clase: rec.sport.hockey'

'CONTENIDO: In  <1qvos8$r78@cl.msu.>, vergolin@euler.lbs.msu.edu (David Vergolini) writes...\n\nThere\'s quite a few Wings fans lurking about here, they just tend\nto be low key and thoughtful rather than woofers.  I suppose every\nfamily must have a Roger Clinton, though.  But remember (to paraphrase\none of my favorite Star Trek lines), "if we adopt the ways of the Leaf\nfans, we are as bad as the Leaf fans".\n\nRon'

'3) Doc similar ID: 7478, similitud coseno: 0.13582073511489307, clase: rec.sport.hockey'

'CONTENIDO: Toronto                          1 1 1--3\nDetroit                          1 4 1--6\nFirst period\n     1, Detroit, Yzerman 1 (Gallant, Ciccarelli) 4:48.\n     2, Toronto, Cullen 1 (Clark, Gill) 10:44.\nSecond period\n     3, Detroit, Sheppard 1 (Probert, Coffey) pp, 5:04.\n     4, Detroit, Burr 1 (Racine) sh, 6:42.\n     5, Detroit, Chiasson 1 (Coffey) pp,11:00.\n     6, Detroit, Howe 1 (Yzerman, Drake) 14:46.\n     7, Toronto, Gilmour 1 (Borschevsky, Ellett) pp, 19:59.\nThird period\n     8, Detroit, Racine 1 (Primeau, Drake) 5:10.\n     9, Toronto, Lefebvre 1 (Cullen, Pearson) 7:45.\n\nDetroit: 6    Power play: 6-2   Special goals:  pp: 2  sh: 1  Total: 3\nScorer            G    A   Pts\n---------------  ---  ---  ---\nBurr               1    0    1\nChiasson           1    0    1\nCiccarelli         0    1    1\nCoffey             0    2    2\nDrake              0    2    2\nGallant            0    1    1\nHowe               1    0    1\nPrimeau            0    1    1\

'4) Doc similar ID: 7308, similitud coseno: 0.13177395037929182, clase: rec.sport.hockey'

"CONTENIDO: Detroit is a very disciplined team.  There's a lot of Europeans\nin Detroit which would make the game fast, so Toronto would have\nto slow the game down, which means drawing penalties, as a last\nresort anyway.  Toronto will be a good team as soon as they get\nmore good players.  Toronto is just an average team, Detroit isn't\nBallard screwed Toronto when he was owner.  Everyone knows that.\nand it's going to take time for Toronto to become a real force.\nI expect Gilmour to be burnt out next year.  He can't pull the\nwhole team forever."

'5) Doc similar ID: 10792, similitud coseno: 0.13073210977865626, clase: rec.sport.baseball'

'CONTENIDO: \n\nThe tribe will be in town from April 16 to the 19th.\nThere are ALWAYS tickets available! (Though they are playing Toronto,\nand many Toronto fans make the trip to Cleveland as it is easier to\nget tickets in Cleveland than in Toronto.  Either way, I seriously\ndoubt they will sell out until the end of the season.)\n'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

Los resultados obtenidos indican que la similitud coseno aplicada a vectores TF-IDF es una herramienta útil para identificar documentos con contenido relacionado al texto de referencia, tanto desde el punto de vista semántico como categórico siempre y cuando exista una alta correlación. 

Cuando los valores de similitud son elevados, es común que los documentos recuperados pertenezcan a la misma clase y compartan términos específicos del tema. No obstante, en situaciones donde la similitud medida es baja, los documentos obtenidos tienden a no guardar una conexión semántica clara ni coincidir en categoría, lo que sugiere que la vectorización no logra reflejar adecuadamente el contenido textual.

### Punto 2

In [89]:
def cls_test (X_train_tfidf, X_test_tfidf):
    clf = MultinomialNB()
    clf.fit(X_train_tfidf, y_train)
    y_pred =  clf.predict(X_test_tfidf)
    display(f"F1 score para MultinomialNB {f1_score(y_test, y_pred, average='macro')}")

    clf = ComplementNB()
    clf.fit(X_train_tfidf, y_train)
    y_pred =  clf.predict(X_test_tfidf)
    display(f"F1 score para ComplementNB {f1_score(y_test, y_pred, average='macro')}")

In [104]:
newsgroups_train = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))
newsgroups_test = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'))

tfidfvect = TfidfVectorizer()
X_train_tfidf = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target
X_test_tfidf = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

display("TF-IDF default")
cls_test (X_train_tfidf, X_test_tfidf)

tfidfvect = TfidfVectorizer(stop_words='english', ngram_range=(1,2), max_df=0.9, min_df=2)
X_train_tfidf = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target
X_test_tfidf = tfidfvect.transform(newsgroups_test.data)
y_test = newsgroups_test.target

display("TF-IDF con parámetros")
cls_test (X_train_tfidf, X_test_tfidf)

'TF-IDF default'

'F1 score para MultinomialNB 0.5854345727938506'

'F1 score para ComplementNB 0.692953349950875'

'TF-IDF con parámetros'

'F1 score para MultinomialNB 0.6503941723196313'

'F1 score para ComplementNB 0.7037173109758078'

Se compararon los modelos MultinomialNB y ComplementNB utilizando representaciones TF-IDF, tanto con la configuración por defecto como con parámetros ajustados (eliminación de stop words en inglés, uso de n-gramas y filtros de frecuencia). 

Esta diferencia se debe a que ComplementNB fue diseñado como una mejora del MultinomialNB para manejar mejor clases desbalanceadas o términos poco representativos, lo que lo vuelve más robusto en configuraciones básicas. Sin embargo, al aplicar una vectorización más cuidada, MultinomialNB mejora notablemente (de 0.585 a 0.650), reduciendo la brecha de desempeño con ComplementNB, que también mejora ligeramente a 0.704.

Estos resultados muestran que ComplementNB es más tolerante a configuraciones simples, mientras que MultinomialNB se beneficia significativamente de una selección adecuada de características.

### Punto 3

In [ ]:
def get_words_cosine_similarity(X_train_T, tfidfvect, word, n):
    topn_dict = {"word": word}
    try:
        word_idx = tfidfvect.vocabulary_[word]
    except:
        print(f"{word} no in vocabulary")
    topn_dict.update({"word_idx": word_idx})
    cossim = cosine_similarity(X_train_T[word_idx], X_train_T)[0]
    topn_dict.update({"word_sim_idx" : np.argsort(cossim)[::-1][1:n+1], "word_sim_cosin" : np.sort(cossim)[::-1][1:n+1]})
    return topn_dict

def print_word_info(idx2word, topn_words):
    for words in topn_words:
        display(f"Palabra referencia {words['word']}, ID:{words['word_idx']}")
        for i in range(len(words['word_sim_idx'])):
            display(f"{i+1}) Palabra similar {idx2word[words['word_sim_idx'][i]]}, ID: {words['word_sim_idx'][i]}, similitud coseno: {words['word_sim_cosin'][i]}")
        display('-'*150)

In [135]:
tfidfvect = TfidfVectorizer(stop_words='english', max_df=0.9, min_df=2) #No voy a utilizar ngrams para que no me busque por pares de palabras
X_train_tfidf = tfidfvect.fit_transform(newsgroups_train.data)
idx2word = {v: k for k,v in tfidfvect.vocabulary_.items()}
y_train = newsgroups_train.target

X_train_T = X_train_tfidf.transpose()
words2test = ['motorcycle', 'cpu', 'football', 'atheist', 'toronto']
topn_words = []
for word in words2test:
    topn_words.append(get_words_cosine_similarity(X_train_T, tfidfvect, word, 5))
print_word_info(idx2word, topn_words)

'Palabra referencia motorcycle, ID:23793'

'1) Palabra similar wheelie, ID: 37637, similitud coseno: 0.404315501394809'

'2) Palabra similar shaft, ID: 31561, similitud coseno: 0.2692545582932047'

'3) Palabra similar learner, ID: 21091, similitud coseno: 0.23783674199976998'

'4) Palabra similar bluntly, ID: 7229, similitud coseno: 0.2302796637099688'

'5) Palabra similar enthusiast, ID: 13800, similitud coseno: 0.2256922997195348'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Palabra referencia cpu, ID:10550'

'1) Palabra similar dislodge, ID: 12367, similitud coseno: 0.356190268602508'

'2) Palabra similar heatsink, ID: 17394, similitud coseno: 0.3086652738160754'

'3) Palabra similar cooler, ID: 10292, similitud coseno: 0.29130050924035983'

'4) Palabra similar bending, ID: 6810, similitud coseno: 0.25026188658596993'

'5) Palabra similar tasking, ID: 34291, similitud coseno: 0.2498389957627389'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Palabra referencia football, ID:15412'

'1) Palabra similar madden, ID: 22094, similitud coseno: 0.34695776542107865'

'2) Palabra similar mich, ID: 23097, similitud coseno: 0.304655737958004'

'3) Palabra similar zabriskie, ID: 38913, similitud coseno: 0.29979462178249405'

'4) Palabra similar basketball, ID: 6554, similitud coseno: 0.2933647593152447'

'5) Palabra similar sportstalk, ID: 32810, similitud coseno: 0.28860827446505277'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Palabra referencia atheist, ID:5889'

'1) Palabra similar atheism, ID: 5888, similitud coseno: 0.33245031466909036'

'2) Palabra similar anarchal, ID: 5086, similitud coseno: 0.31034751316326226'

'3) Palabra similar atheists, ID: 5891, similitud coseno: 0.2947785138421634'

'4) Palabra similar doubtless, ID: 12747, similitud coseno: 0.29430041416345676'

'5) Palabra similar techno, ID: 34379, similitud coseno: 0.2861931096814526'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

'Palabra referencia toronto, ID:35041'

'1) Palabra similar detroit, ID: 11930, similitud coseno: 0.3812332304163351'

'2) Palabra similar gilmour, ID: 16314, similitud coseno: 0.2479111071502925'

'3) Palabra similar chicago, ID: 8879, similitud coseno: 0.24710485460503068'

'4) Palabra similar eecg, ID: 13301, similitud coseno: 0.19442096527109395'

'5) Palabra similar disciplined, ID: 12269, similitud coseno: 0.1834678555585414'

'------------------------------------------------------------------------------------------------------------------------------------------------------'

Transponer la matriz TF-IDF de documentos genera una matriz que representa términos en función de los documentos en los que aparecen. El significado de cada término queda definido por su distribución a lo largo del corpus, lo cual está limitado por la cantidad y la categoría de los documentos disponibles.

En los ejemplos seleccionados manualmente se observa que términos como 'motorcycle' están fuertemente asociados a 'wheelie' y 'shaft', o que 'cpu' se relaciona con 'heatsink', 'cooler' o 'tasking'. De forma similar, Toronto muestra correlación con otras ciudades como Detroit o Chicago. Sin embargo, el método no es completamente preciso, ya que algunas asociaciones carecen de sentido aparente, posiblemente porque las relaciones entre palabras se manifiestan a nivel de documento y no de coocurrencia directa.

TF-IDF este método no considera contexto, orden de palabras ni sintaxis. Para mejorar la representación semántica de las palabras, se recurre a técnicas más avanzadas como los word embeddings, que capturan relaciones contextuales de forma más robusta.